### PR TITLE
MM-48793, MM-48137: improved bot_joined|left_channel|team

### PR DESCRIPTION
**Pull request by levb**
#### Summary
The `bot_joined|left_...` notifications had several issues making them dysfunctional for the use-cases. This PR adds new subjects `self_joined|left_...` and also provides backwards compatibility for the `bot_` subjects. 1/5 I'd prefer to remove the bot-specific subjects altogether, but not sure if it'd break any existing apps.

- Added `self_joined|left_channel|team` subscription subjects, available to all users not just the bot. These are globally scoped since they are already limited to the subscribing user.
- Changed Expand to bypass permission checks for these events since they are limited to the member of the team|channel.
- Replaced `server/mmclient` with `server/proxy/ExpandGetter`
- Deprecated `bot_joined|left_channel|team`
  - Previously stored (global?) subscriptions work.
  - New `bot_...` subscriptions are stored as `self_...`
- A small logging improvement in `goapp`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48137
https://mattermost.atlassian.net/browse/MM-48793

### CODE REVIEW

File: apps/subscription.go [hunk 1]
```diff
@@ -128,7 +127,9 @@ func (e Event) validate(appendTo error) error {
 	// Globally scoped, must not contain any extra qualifiers.
 	case SubjectUserCreated,
 		SubjectBotJoinedTeam,
-		SubjectBotLeftTeam /*, SubjectBotMentioned*/ :
+		SubjectBotLeftTeam,
```
Comment (levb) [inline apps/subscription.go:130]: reminder: Change comments in line 52

File: apps/subscription.go [hunk 2]
```diff
@@ -115,8 +115,7 @@ func (sub Subscription) Validate() error {
 }
 
 func (e Event) Validate() error {
-	var result error
-	return e.validate(result)
+	return e.validate(nil)
```
Comment (javaguirre) [inline apps/subscription.go:118]: What does this do?

File: apps/subscription.go [hunk 3]
```diff
@@ -139,24 +140,22 @@ func (e Event) validate(appendTo error) error {
 	// Team scoped, require TeamID, no ChannelID
 	case SubjectUserJoinedTeam,
 		SubjectUserLeftTeam,
-		SubjectBotJoinedChannel,
-		SubjectBotLeftChannel,
 		SubjectChannelCreated:
 		if e.TeamID == "" {
-			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a team; teamID must not be empty", e.Subject))
+			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a team; team_id must not be empty", e.Subject))
```
Comment (javaguirre) [inline apps/subscription.go:145]: is this an error? I see the element is called `TeamID`, not `team_id`? In line 140 is also called that way.

File: apps/subscription.go [hunk 4]
```diff
@@ -139,24 +140,22 @@ func (e Event) validate(appendTo error) error {
...
 	// Channel scoped, require ChannelID, no TeamID
 	case SubjectUserJoinedChannel,
 		SubjectUserLeftChannel /*, SubjectPostCreated */ :
 		if e.TeamID != "" {
-			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a channel; teamID must be empty", e.Subject))
+			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a channel; team_id must be empty", e.Subject))
 		}
 		if e.ChannelID == "" {
-			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a channel; ChannelID must not be empty", e.Subject))
+			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is scoped to a channel; channel_id must not be empty", e.Subject))
```
Comment (javaguirre) [inline apps/subscription.go:158]: Idem above

**Review by javaguirre — COMMENTED** [review]
I also have some questions regarding naming. 👍

File: apps/subscription.go (hunk 3 above)
Comment (levb) [reply apps/subscription.go:145]: The JSON is lowercase.

**Review by javaguirre — APPROVED** [review]

**Review by mickmister — APPROVED** [review]

**Comment by levb** [comment]
The testing of this change should be covered with automated tests. @mickmister @javaguirre if you find them satisfactory, do you mind removing "QA Review" label?

**Comment by levb** [comment]
1/5 I am not sure why do we need this (backend-only) change covered in the test app. If the automated tests are not adequate, they can (should?) be augmented. But since it is possible to cover the functionality entirely with automated tests, I would advocate for that.

The test app is already byzantine-complicated, and not documented.

**Comment by javaguirre** [comment]
> 1/5 I am not sure why do we need this (backend-only) change covered in the test app. If the automated tests are not adequate, they can (should?) be augmented. But since it is possible to cover the functionality entirely with automated tests, I would advocate for that.
> 
> The test app is already byzantine-complicated, and not documented.

In that case, let's review the test app later. I think some [parts of the documentation don't reflect current reality anyway](https://developers.mattermost.com/integrate/apps/authentication/app-to-third-party/), so it would be a good exercise.

**Comment by mickmister** [comment]
> 1/5 I am not sure why do we need this (backend-only) change covered in the test app

The test app currently that it must provide a team id in this subscription, and I was suggesting that it follows suit with this change. If subscribe still works when this team id is supplied then I suppose that's fine, but if this change breaks the test app, then it should be updated.

File: server/appservices/service.go [hunk 5]
```diff
@@ -12,7 +12,7 @@ import (
 type Service interface {
 	// Subscriptions
 
-	Subscribe(*incoming.Request, apps.Subscription) error
+	Subscribe(_ *incoming.Request, _ apps.Subscription, testBypassBotSubjectMapping bool) error
```
Comment (levb) [inline server/appservices/service.go:15]: to store unmapped legacy subscriptions, for testing.

**Review by mickmister — APPROVED** [review]
LGTM :+1:

**Review by javaguirre — APPROVED** [review]

File: server/store/subscriptions.go [hunk 6]
```diff
@@ -42,8 +42,15 @@ func subsKey(e apps.Event) (string, error) {
 	idSuffix := ""
 	switch e.Subject {
 	case apps.SubjectUserCreated,
+		// apps.SubjectSelfMentioned,
```
[the remark this answers is not in the document]
Comment (levb) [reply server/store/subscriptions.go:45]: Just a reminder for the future. It was (sort of) there before for the bot subject.

**Review by asatkinson — COMMENTED** [review]
@DHaussermann @hanzei can you please have a review of this.  Peritus is waiting for this fix

**Comment by levb** [comment]
Closing in favor of #443

**Pull request closed without merging**